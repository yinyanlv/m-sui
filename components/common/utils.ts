export function debounce(fn: any, time: number): (...args: any[]) => any {

    let timer: ReturnType<typeof setTimeout>;

    return (...args: any[]): any => {
       
        if (timer) {
            clearTimeout(timer);
        }

        timer = setTimeout(() => {
            fn && fn(...args);
        }, time);
    };
}