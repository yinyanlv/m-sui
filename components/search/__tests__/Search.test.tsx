import React from 'react';
import { shallow, mount } from 'enzyme';
import { Search } from '..';
import { sleep } from '../../../test/utils';


describe('Search', () => {
    const onSearch = jest.fn();
    const placeholder = '请输入';

    it('should render', () => {
        const wrap = shallow(<Search placeholder={placeholder} onSearch={onSearch} />);

        expect(wrap.render()).toMatchSnapshot();
    });

    it('should render without throw', () => {
        const fn = () => {
            shallow(<Search placeholder={placeholder} onSearch={onSearch} />);
        };

        expect(fn).not.toThrow();
    });

    it('placeholder should be set', () => {
        const wrap = shallow(<Search placeholder={placeholder} onSearch={onSearch} />);
        expect(wrap.find('span.placeholder').text().trim()).toEqual(placeholder);
    });

    it('should trigger onSearch when click search button', () => {
        const onSearch = jest.fn();
        const wrap = mount(<Search placeholder={placeholder} onSearch={onSearch} />);

        wrap.find('input').simulate('focus');
        wrap.find('.btn').simulate('click');

        expect(onSearch).toBeCalledTimes(1);
        expect(onSearch).toBeCalledWith('');
    });

    it('should manual set value', () => {
        const ref: any = React.createRef();
        const wrap = mount(<Search placeholder={placeholder} onSearch={onSearch} ref={ref} />)

        ref.current.setValue('test');

        const input: any = wrap.find('input').getDOMNode();

        expect(input.value).toEqual('test');
    });

    it('should trigger onChange', async () => {
        const ref: any = React.createRef();
        const onChange = jest.fn();

        const wrap = mount(<Search placeholder={placeholder} onSearch={onSearch} onChange={onChange} debounceTime={500} ref={ref} />);

        ref.current.setValue('test');
        wrap.find('input').simulate('change');
        expect(onChange).toBeCalledTimes(0);
        // test debounce
        await sleep(1000);
        expect(onChange).toBeCalledTimes(1);
        expect(onChange).lastCalledWith('test');
    });

    it('should preventEvent when submit form', () => {
        const onSearch = jest.fn();
        let prevented = false;
        const wrap = mount(<Search placeholder={placeholder} onSearch={onSearch} />);

        wrap.find('input').simulate('submit', {
            preventDefault: () => {
                prevented = true;
            }
        });

        expect(prevented).toBeTruthy();
    });
});