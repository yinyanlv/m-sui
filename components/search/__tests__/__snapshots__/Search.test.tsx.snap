// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render 1`] = `
initialize {
  "0": Object {
    "attribs": Object {},
    "children": Array [
      Object {
        "attribs": Object {
          "class": "input-wrap",
        },
        "children": Array [
          Object {
            "attribs": Object {
              "action": "/",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "autocomplete": "off",
                  "type": "search",
                },
                "children": Array [],
                "name": "input",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": null,
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "autocomplete": undefined,
                  "type": undefined,
                },
                "x-attribsPrefix": Object {
                  "autocomplete": undefined,
                  "type": undefined,
                },
              },
            ],
            "name": "form",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": Object {
              "attribs": Object {
                "class": "placeholder",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "class": "iconfont icon-magnifier",
                  },
                  "children": Array [],
                  "name": "i",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": Object {
                    "data": " 请输入",
                    "next": null,
                    "parent": [Circular],
                    "prev": [Circular],
                    "type": "text",
                  },
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                Object {
                  "data": " 请输入",
                  "next": null,
                  "parent": [Circular],
                  "prev": Object {
                    "attribs": Object {
                      "class": "iconfont icon-magnifier",
                    },
                    "children": Array [],
                    "name": "i",
                    "namespace": "http://www.w3.org/1999/xhtml",
                    "next": [Circular],
                    "parent": [Circular],
                    "prev": null,
                    "type": "tag",
                    "x-attribsNamespace": Object {
                      "class": undefined,
                    },
                    "x-attribsPrefix": Object {
                      "class": undefined,
                    },
                  },
                  "type": "text",
                },
              ],
              "name": "span",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": null,
              "parent": [Circular],
              "prev": [Circular],
              "type": "tag",
              "x-attribsNamespace": Object {
                "class": undefined,
              },
              "x-attribsPrefix": Object {
                "class": undefined,
              },
            },
            "parent": [Circular],
            "prev": null,
            "type": "tag",
            "x-attribsNamespace": Object {
              "action": undefined,
            },
            "x-attribsPrefix": Object {
              "action": undefined,
            },
          },
          Object {
            "attribs": Object {
              "class": "placeholder",
            },
            "children": Array [
              Object {
                "attribs": Object {
                  "class": "iconfont icon-magnifier",
                },
                "children": Array [],
                "name": "i",
                "namespace": "http://www.w3.org/1999/xhtml",
                "next": Object {
                  "data": " 请输入",
                  "next": null,
                  "parent": [Circular],
                  "prev": [Circular],
                  "type": "text",
                },
                "parent": [Circular],
                "prev": null,
                "type": "tag",
                "x-attribsNamespace": Object {
                  "class": undefined,
                },
                "x-attribsPrefix": Object {
                  "class": undefined,
                },
              },
              Object {
                "data": " 请输入",
                "next": null,
                "parent": [Circular],
                "prev": Object {
                  "attribs": Object {
                    "class": "iconfont icon-magnifier",
                  },
                  "children": Array [],
                  "name": "i",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": [Circular],
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "class": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "class": undefined,
                  },
                },
                "type": "text",
              },
            ],
            "name": "span",
            "namespace": "http://www.w3.org/1999/xhtml",
            "next": null,
            "parent": [Circular],
            "prev": Object {
              "attribs": Object {
                "action": "/",
              },
              "children": Array [
                Object {
                  "attribs": Object {
                    "autocomplete": "off",
                    "type": "search",
                  },
                  "children": Array [],
                  "name": "input",
                  "namespace": "http://www.w3.org/1999/xhtml",
                  "next": null,
                  "parent": [Circular],
                  "prev": null,
                  "type": "tag",
                  "x-attribsNamespace": Object {
                    "autocomplete": undefined,
                    "type": undefined,
                  },
                  "x-attribsPrefix": Object {
                    "autocomplete": undefined,
                    "type": undefined,
                  },
                },
              ],
              "name": "form",
              "namespace": "http://www.w3.org/1999/xhtml",
              "next": [Circular],
              "parent": [Circular],
              "prev": null,
              "type": "tag",
              "x-attribsNamespace": Object {
                "action": undefined,
              },
              "x-attribsPrefix": Object {
                "action": undefined,
              },
            },
            "type": "tag",
            "x-attribsNamespace": Object {
              "class": undefined,
            },
            "x-attribsPrefix": Object {
              "class": undefined,
            },
          },
        ],
        "name": "div",
        "namespace": "http://www.w3.org/1999/xhtml",
        "next": null,
        "parent": [Circular],
        "prev": null,
        "type": "tag",
        "x-attribsNamespace": Object {
          "class": undefined,
        },
        "x-attribsPrefix": Object {
          "class": undefined,
        },
      },
    ],
    "name": "div",
    "namespace": "http://www.w3.org/1999/xhtml",
    "next": null,
    "parent": null,
    "prev": null,
    "root": Object {
      "attribs": Object {},
      "children": Array [
        [Circular],
      ],
      "name": "root",
      "namespace": "http://www.w3.org/1999/xhtml",
      "next": null,
      "parent": null,
      "prev": null,
      "type": "root",
      "x-attribsNamespace": Object {},
      "x-attribsPrefix": Object {},
    },
    "type": "tag",
    "x-attribsNamespace": Object {},
    "x-attribsPrefix": Object {},
  },
  "_root": [Circular],
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "normalizeWhitespace": false,
    "withDomLvl1": true,
    "xml": false,
  },
}
`;
